diff --git a/cluster.go b/cluster.go
index c3bf8b0..17a0261 100644
--- a/gopkg.in/mgo.v2/cluster.go
+++ b/gopkg.in/mgo.v2/cluster.go
@@ -646,6 +646,10 @@ func (cluster *mongoCluster) AcquireSocket(mode Mode, slaveOk bool, syncTimeout
 				cluster.syncServers()
 				time.Sleep(100 * time.Millisecond)
 				continue
+			} else {
+				server.Lock()
+				server.abended = false
+				server.Unlock()
 			}
 		}
 		return s, nil
